// azurite/dist/src/queue/QueueEnvironment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const args_1 = tslib_1.__importDefault(require("args"));
const constants_1 = require("./utils/constants");
args_1.default
    .option(["", "queueHost"], "Optional. Customize listening address for queue", constants_1.DEFAULT_QUEUE_SERVER_HOST_NAME)
    .option(["", "queuePort"], "Optional. Customize listening port for queue", constants_1.DEFAULT_QUEUE_LISTENING_PORT)
    .option(["l", "location"], "Optional. Use an existing folder as workspace path, default is current working directory", process.cwd())
    .option(["s", "silent"], "Optional. Disable access log displayed in console")
    .option(["L", "loose"], "Optional. Enable loose mode which ignores unsupported headers and parameters")
    .option(["", "disableProductStyleUrl"], "Optional. DDisable getting account name from the host of request Uri, always get account name from the first path segment of request Uri.")
    .option(["", "skipApiVersionCheck"], "Optional. Skip the request API version check, request with all Api versions will be allowed")
    .option(["", "oauth"], 'Optional. OAuth level. Candidate values: "basic"')
    .option(["", "cert"], "Optional. Path to certificate file")
    .option(["", "key"], "Optional. Path to certificate key .pem file")
    .option(["", "pwd"], "Optional. Password for .pfx file")
    .option(["d", "debug"], "Optional. Enable debug log by providing a valid local file path as log destination");
args_1.default.config.name = "azurite-queue";
class QueueEnvironment {
    constructor() {
        this.flags = args_1.default.parse(process.argv);
    }
    queueHost() {
        return this.flags.queueHost;
    }
    queuePort() {
        return this.flags.queuePort;
    }
    async location() {
        return this.flags.location || process.cwd();
    }
    silent() {
        if (this.flags.silent !== undefined) {
            return true;
        }
        return false;
    }
    loose() {
        if (this.flags.loose !== undefined) {
            return true;
        }
        // default is false which will block not supported APIs, headers and parameters
        return false;
    }
    skipApiVersionCheck() {
        if (this.flags.skipApiVersionCheck !== undefined) {
            return true;
        }
        // default is false which will check API veresion
        return false;
    }
    cert() {
        return this.flags.cert;
    }
    key() {
        return this.flags.key;
    }
    pwd() {
        return this.flags.pwd;
    }
    oauth() {
        return this.flags.oauth;
    }
    disableProductStyleUrl() {
        if (this.flags.disableProductStyleUrl !== undefined) {
            return true;
        }
        // default is false which will try to get account name from request Uri hostname
        return false;
    }
    async debug() {
        if (typeof this.flags.debug === "string") {
            // Enable debug log to file
            return this.flags.debug;
        }
        if (this.flags.debug === true) {
            throw RangeError(`Must provide a debug log file path for parameter -d or --debug`);
        }
        // By default disable debug log
    }
}
exports.default = QueueEnvironment;
